### PR TITLE
Enhancement unify tool creation

### PR DESCRIPTION
**Pull request by DSWallach**
# Pull Request Template for MMEDS

## What has changed
Include each file that was changed, how it was changed (in broad terms), and why it was changed.
Watcher now checks the mongo documents for the status of the processes it creates.
Tool setup now happens within the tool itsself 

## Checklist of pre-requisites
-   [x] Does the code run?  
-   [x] Does the code follow the repository style?  
-   [x] Is the code tested?  

## How to use the feature
Example: To do X run new function Y with parameters A, B, and C which gives result Z.

## How the design has changed
Example: Functions A, B, and C are now all methods of Class D.

### CODE REVIEW

File: mmeds/documents.py [hunk 1]
```diff
@@ -35,43 +36,69 @@ class MMEDSDoc(men.Document):
...
-                # Otherwise just write the value
-                else:
-                    f.write('{}\t{}\n'.format(key, file_path))
-        with open(DOCUMENT_LOG, 'a') as f:
-            f.write('-\t'.join([str(type(self)), self.owner, 'Upload', 'Finished', self.path, self.access_code]) + '\n')
-        super().save()
+    def save(self, **kwargs):
+        super().save(**kwargs)
+        if self.path is not None:
+            with open(str(Path(self.path) / 'file_index.tsv'), 'w') as f:
```
Comment (mstapylton) [inline mmeds/documents.py]: A little nit-picky, but I don't like the use of f as a name for a file object. My preference would be for something more descriptive.

File: mmeds/documents.py [hunk 2]
```diff
@@ -35,43 +36,69 @@ class MMEDSDoc(men.Document):
...
+                    # Skip non existent files
+                    if file_path is None:
+                        continue
+                    # If it's a key for an analysis point to the file index for that analysis
+                    elif isinstance(file_path, dict):
+                        f.write('{}\t{}\n'.format(key, Path(self.path) / key / 'file_index.tsv'))
+                    # Otherwise just write the value
+                    else:
+                        f.write('{}\t{}\n'.format(key, file_path))
+            with open(DOCUMENT_LOG, 'a') as f:
```
Comment (mstapylton) [inline mmeds/documents.py]: Again, don't like the use of f as the name of the file object.

File: mmeds/spawn.py [hunk 3]
```diff
@@ -109,30 +100,47 @@ def __init__(self, queue, pipe, parent_pid, testing=False):
...
 
-    def add_process(self, ptype, process):
+    def add_process(self, ptype, process_code):
         """ Add an analysis process to the list of processes. """
-        self.running_processes.append(process)
+        self.running_processes.append(process_code)
         self.write_running_processes()
 
     def check_processes(self):
         still_running = []
```
Comment (mstapylton) [inline mmeds/spawn.py]: docstring?

File: mmeds/spawn.py [hunk 4]
```diff
@@ -154,17 +162,66 @@ def log_processes(self):
...
-
     def any_running(self, ptype):
         """ Returns true if there is a process running """
         return bool(self.running_processes['ptype'])
 
     def get_processes(self):
         return self.running_processes, self.processes
 
+    def handle_analysis(self, process):
+        ptype, user, access_code, tool_type, analysis_type, config, kill_stage = process
```
Comment (mstapylton) [inline mmeds/spawn.py]: docstring?

File: mmeds/spawn.py [hunk 5]
```diff
@@ -154,17 +162,66 @@ def log_processes(self):
...
+        p.start()
+        sleep(1)
+        with Database(testing=self.testing, owner=user) as db:
+            doc = db.get_doc(p.access_code)
+        self.pipe.send(doc.get_info())
+        # Add it to the list of analysis processes
+        self.add_process(ptype, p.access_code)
+
+    def handle_upload(self, process, current_upload):
+        # Check that there isn't another process currently uploading
```
Comment (mstapylton) [inline mmeds/spawn.py]: docstring?

File: mmeds/spawn.py [hunk 6]
```diff
@@ -154,17 +162,66 @@ def log_processes(self):
...
+                doc = db.get_doc(p.access_code, False)
+            self.pipe.send(doc.get_info())
+            self.started.append(p.access_code)
+            current_upload = p
+            if self.testing:
+                p.join()
+            return current_upload
+
+    def handle_restart(self, process):
+        ptype, user, analysis_code, restart_stage, kill_stage = process
```
Comment (mstapylton) [inline mmeds/spawn.py]: docstring?

File: mmeds/spawn.py [hunk 7]
```diff
@@ -109,30 +100,47 @@ def __init__(self, queue, pipe, parent_pid, testing=False):
...
-            else:
-                self.processes.append(process)
-                # Send the exitcode of the process
-                self.pipe.send(process.exitcode)
+        with Database(testing=self.testing) as db:
+            # For each type of process 'upload', 'analysis', 'etc'
+            # Check each process is still alive
+            while self.running_processes:
+                process_code = self.running_processes.pop()
+                while True:
```
Comment (mstapylton) [inline mmeds/spawn.py]: what's the purpose of having
```while True:``` 
... 
```break```
to just run the code once?

File: mmeds/documents.py (hunk 1 above)
Comment (DSWallach) [reply mmeds/documents.py]: In general I agree single letter variables are bad, but `f` is a standard for files in python. A bit like `i` for iterating. It's even used in the documentation. https://docs.python.org/3/tutorial/inputoutput.html That being said, I want to replaces this all with a yaml file, so there will be many fewer lines of code where `f` is defined

Comment (mstapylton) [reply mmeds/documents.py]: I think it depends on how ambiguous ```f``` is in the given context. For instance, if you had a function that just opened a file, I think using f to represent a generic file makes sense. By contrast, your ```f``` file here seems to reference a specific file ```../../file_index.tsv``` which is referenced a few times. However, if you're going to change the code down the road anyway then I think it's fine.

**Pull request closed without merging**